src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs

```diff
@@ -1442,5 +1442,141 @@ public void User_WrappedFunc_Throw_Exceptions()
+            driver = driver.RunGenerators(compilation);
+            Assert.Equal(1, compilationsCalledFor.Count);
+            Assert.Equal(compilation, compilationsCalledFor[0]);
+        }
+
+        [Fact]
+        public void IncrementalGenerator_Runs_Only_For_Changed_Inputs()
+        {
+            var source = @"
+class C { }
+";
+            var parseOptions = TestOptions.Regular.WithLanguageVersion(LanguageVersion.Preview);
+            Compilation compilation = CreateCompilation(source, options: TestOptions.DebugDll, parseOptions: parseOptions);
+            compilation.VerifyDiagnostics();
+
+            Assert.Single(compilation.SyntaxTrees);
+
+            var text1 = new InMemoryAdditionalText("Text1", "content1");
+            var text2 = new InMemoryAdditionalText("Text2", "content2");
+
+            List<Compilation> compilationsCalledFor = new List<Compilation>();
+            List<AdditionalText> textsCalledFor = new List<AdditionalText>();
+
+            var generator = new IncrementalGeneratorWrapper(new PipelineCallbackGenerator(ctx =>
+            {
+                ctx.RegisterOutput(ctx.Sources.Compilation.GenerateSource((spc, c) =>
+                {
+                    compilationsCalledFor.Add(c);
+                }));
+
+                ctx.RegisterOutput(ctx.Sources.AdditionalTexts.GenerateSource((spc, c) =>
+                {
+                    textsCalledFor.Add(c);
+                }));
+            }));
+
+
+            // run the generator once, and check it was passed the compilation
+            GeneratorDriver driver = CSharpGeneratorDriver.Create(new ISourceGenerator[] { generator }, additionalTexts: new[] { text1 }, parseOptions: parseOptions);
+            driver = driver.RunGenerators(compilation);
+            Assert.Equal(1, compilationsCalledFor.Count);
+            Assert.Equal(compilation, compilationsCalledFor[0]);
+            Assert.Equal(1, textsCalledFor.Count);
+            Assert.Equal(text1, textsCalledFor[0]);
+
+            // clear the results, add an aditional text, but keep the compilation the same
+            compilationsCalledFor.Clear();
+            textsCalledFor.Clear();
+            driver = driver.AddAdditionalTexts(ImmutableArray.Create<AdditionalText>(text2));
+            driver = driver.RunGenerators(compilation);
+            Assert.Equal(0, compilationsCalledFor.Count);
+            Assert.Equal(1, textsCalledFor.Count);
+            Assert.Equal(text2, textsCalledFor[0]);
+
+            // now edit the compilation
+            compilationsCalledFor.Clear();
+            textsCalledFor.Clear();
+            var newCompilation = compilation.WithOptions(compilation.Options.WithModuleName("newComp"));
+            driver = driver.RunGenerators(newCompilation);
+            Assert.Equal(1, compilationsCalledFor.Count);
+            Assert.Equal(newCompilation, compilationsCalledFor[0]);
+            Assert.Equal(0, textsCalledFor.Count);
```

Comment:
>(0, textsCalledFor.Count) [](http://example.com/codeflow?start=24&length=25)

Why wasn't the generator called for each text file if CompilationOptions changed? How would a generator be written where CompilationOptions might affect each generated source file?

Reply:
Because the above generator essentially has two pipelines, one transforming the compilation, and the other transforming the additional texts. The additional text pipeline can't access the compilation options, so we know by construction that its correct to not call it. 

If the author joined the additional texts input to the compilation options source, we'd call it when either changed.

---
In reply to: [632179392](https://github.com/dotnet/roslyn/pull/53383#discussion_r632179392) [](ancestors = 632179392)